File: codeforces/611/C/C.cpp
#include <iostream>
#include <map>
#include <utility>
#include <cstdio>

using namespace std;

const int MAX_HW = 520;

int h, w;
char grid[MAX_HW][MAX_HW];

bool isHGood(int r, int c) {
  if (c == w) {
    return false;
  } else {
    return grid[r-1][c-1] == '.' && grid[r-1][c] == '.';
  }
}

bool isVGood(int r, int c) {
  if (r == h) {
    return false;
  } else {
    return grid[r-1][c-1] == '.' && grid[r][c-1] == '.';
  }
}

int numHGoodMemo[MAX_HW][MAX_HW];

int numHGood(int r, int c) {

  if (numHGoodMemo[r][c] != -1)
    return numHGoodMemo[r][c];

  if (r <= 1) {
    int ans = 0;
    for (int x=1; x<=r; x++)
      for (int y=1; y<=c; y++)
        if (isHGood(x, y)) ans += 1;
    numHGoodMemo[r][c] = ans;
  } else {
    int ans = numHGood(r-1, c);
    for (int y=1; y<= c; y++) {
      if (isHGood(r, y)) ans += 1;
    }
    numHGoodMemo[r][c] = ans;
  }
  return numHGoodMemo[r][c];

}

int numVGoodMemo[MAX_HW][MAX_HW];
int numVGood(int r, int c) {
  if (numVGoodMemo[r][c] != -1)
    return numVGoodMemo[r][c];

  if (r <= 1) {
    int ans = 0;
    for (int x=1; x<=r; x++)
      for (int y=1; y<=c; y++)
        if (isVGood(x, y)) ans += 1;

    numVGoodMemo[r][c] = ans;
  } else {
    int ans = numVGood(r-1, c);
    for (int y=1; y<= c; y++) {
      if (isVGood(r, y)) ans += 1;
    }
    numVGoodMemo[r][c] = ans;
  }
  return numVGoodMemo[r][c];
}

int numHGood4(int r1, int c1, int r2, int c2) {
  return numHGood(r2, c2) - numHGood(r1-1, c2) - numHGood(r2, c1-1) + numHGood(r1-1, c1-1);
}

int numVGood4(int r1, int c1, int r2, int c2) {
  return numVGood(r2, c2) - numVGood(r1-1, c2) - numVGood(r2, c1-1) + numVGood(r1-1, c1-1);
}

int numDominoes(int r1, int c1, int r2, int c2) {
  return numHGood4(r1, c1, r2, c2-1) + numVGood4(r1, c1, r2-1, c2);
}

int main () {
  scanf("%d %d\n", &h, &w);

  for (int i=0; i<MAX_HW; i++) {
    for (int j=0; j<MAX_HW; j++) {
      numHGoodMemo[i][j] = -1;
      numVGoodMemo[i][j] = -1;
    }
  }

  for (int i=0; i<h; i++) {
    for (int j=0; j<w; j++) {
      scanf("%c", &grid[i][j]);
    }
    scanf("\n");
  }

  int q;
  scanf("%d\n", &q);

  for (int i=0; i<q; i++) {
    int r1, c1, r2, c2;
    scanf("%d %d %d %d\n", &r1, &c1, &r2, &c2);
    printf("%d\n", numDominoes(r1, c1, r2, c2));
  }

}

File: codeforces/611/C/C.scala
import scala.collection.mutable.ArrayBuffer
import scala.collection.mutable.Map
import java.util.Scanner

object C {

  def main(args: Array[String]): Unit = {

    val sc = new Scanner(System.in)

    val h = sc.nextInt()
    val w = sc.nextInt()

    sc.nextLine()

    val grid = ArrayBuffer[String]();

    (0 until h).foreach((i) => {
      grid += sc.nextLine()
    })

    def isHGood(r:Int, c:Int) = {
      if (c == w)
        false
      else
        grid(r-1)(c-1) == '.' && grid(r-1)(c) == '.'
    }

    def isVGood(r: Int, c:Int) = {
      if (r == h)
        false
      else
        grid(r-1)(c-1) == '.' && grid(r)(c-1) == '.'
    }

    var numHGoodMemo = Map[(Int, Int), Int]()

    def numHGood(r: Int, c: Int):Int = {

      val key = (r, c)
      if (numHGoodMemo.isDefinedAt(key)) {
        numHGoodMemo.get(key).get
      } else if (r <= 1) {
        var ans = 0
        (1 to r).foreach(x => {
          (1 to c).foreach(y => {
            if (isHGood(x, y)) ans += 1
          })
        })
        numHGoodMemo.update(key, ans)
        ans
      } else {
        var ans = numHGood(r-1, c);
        (1 to c).foreach(y => {
          if (isHGood(r, y)) ans += 1
        })
        numHGoodMemo.update(key, ans)
        ans
      }
    }

    var numVGoodMemo = Map[(Int, Int), Int]()

    def numVGood(r: Int, c: Int):Int = {

      val key = (r, c)
      if (numVGoodMemo.isDefinedAt(key)) {
        numVGoodMemo.get(key).get
      } else if (r <= 1) {
        var ans = 0
        (1 to r).foreach(x => {
          (1 to c).foreach(y => {
            if (isVGood(x, y)) ans += 1
          })
        })
        numVGoodMemo.update(key, ans)
        ans
      } else {
        var ans = numVGood(r-1, c)
        (1 to c).foreach(y => {
          if (isVGood(r, y)) ans += 1
        })
        numVGoodMemo.update(key, ans)
        ans
      }


    }

    def numHGood4(r1:Int, c1:Int, r2:Int, c2:Int):Int = {
      numHGood(r2, c2) - numHGood(r1-1, c2) - numHGood(r2, c1-1) + numHGood(r1-1, c1-1)
    }

    def numVGood4(r1:Int, c1:Int, r2:Int, c2:Int):Int = {
      numVGood(r2, c2) - numVGood(r1-1, c2) - numVGood(r2, c1-1) + numVGood(r1-1, c1-1)
    }

    def numDominoes(r1:Int, c1:Int, r2:Int, c2:Int):Int = {
      numHGood4(r1, c1, r2, c2-1) + numVGood4(r1, c1, r2-1, c2)
    }

    val q = sc.nextInt()

    (0 until q).foreach(_ => {
      val r1 = sc.nextInt()
      val c1 = sc.nextInt()
      val r2 = sc.nextInt()
      val c2 = sc.nextInt()
      println(numDominoes(r1, c1, r2, c2))
    })

  }

}
